### ISSUE TITLE
Error when using always_issue_new_refresh_token and Mongo storage

### ISSUE THREAD

**Issue by weckx** [zfcampus/zf-oauth2#118]
I'm trying to set the always_issue_new_refresh_token option to true in my zf-oauth2 configuration:

```php
    'zf-oauth2' => [
        'options' => [
            'always_issue_new_refresh_token' => true,
        ]
    ],
```

But I get the following error:

```
Uncaught exception 'MongoConnectionException' with message '- Found unknown connection string option 'always_issue_new_refresh_token' with value '1'' 
```

It seems that the _options_ parameters are passed as the second argument of MongoClient, that behavior was updated with the PR#64. I couldn't find any workaround for this config, so I believe there should be an additional parameter inside options to pass the MongoClient specific configurations.

**Comment by codeaid** [comment]
Have the same issue. Looks like the whole `options` array is being passed to `MongoClient` on line 200 in `./zfcampus/zf-mvc-auth/src/Factory/OAuth2ServerFactory.php`.

**Comment by codeaid** [comment]
Update - I managed to get it to work by updating the `OAuth2ServerFactory::marshalOptions()` method as follows:

```php
private static function marshalOptions(array $config)
{
    // merge nested options with the root options
    $options = array_key_exists('options', $config) ? $config['options'] : array();
    return array_replace_recursive($config, $options);
}
```

An alternative is to only allow specifying valid keys and remove the ones that the `Server` class does not expect:

```php
private static function marshalOptions(array $config)
{
    $defaults = [
        'access_lifetime'                   => 3600,
        'allow_credentials_in_request_body' => true,
        'allow_implicit'                    => false,
        'allow_public_clients'              => true,
        'always_issue_new_refresh_token'    => false,
        'enforce_state'                     => true,
        'id_lifetime'                       => 3600,
        'require_exact_redirect_uri'        => true,
        'store_encrypted_token_string'      => true,
        'token_bearer_header_name'          => 'Bearer',
        'token_param_name'                  => 'access_token',
        'unset_refresh_token_after_use'     => true,
        'use_jwt_access_tokens'             => false,
        'use_openid_connect'                => false,
        'www_realm'                         => 'Service',
    ];

    return array_intersect_key($config, $defaults);
}
```

Defaults are copied from [OAuth2\Server::__construct()](https://github.com/bshaffer/oauth2-server-php/blob/master/src/OAuth2/Server.php#L110-L126). Not sure if it's a good idea to do this as the library maintainer could add a new property meaning the `zf-oauth2` has to be then modified to support it.

**Comment by weckx** [comment]
My workaround was separating the mongo options to another key in the config, eg: mongo_options, but I'm also not sure if that's the best option. Any word from the mainteners? @weierophinney ?

**Comment by matwright** [comment]
This issue can be resolved entirely by configuration.

https://github.com/zfcampus/zf-mvc-auth/blob/master/src/Factory/OAuth2ServerFactory.php#L183

```php
 $dbLocatorName = isset($config['locator_name'])
            ? $config['locator_name']
            : 'MongoDB';
        if ($services->has($dbLocatorName)) {
            return $services->get($dbLocatorName);
        }
```

as you can see above the createMongoDatabase method first checks if you have provided an existing mongo instance (as a service name) in your config. If so, then that is used and the options are never passed to the MongoClient where your exception is being thrown.

As you are using zf-oauth2 you must already have the DB adapter configured, hence you can simply provide this adapter as your 'locator_name' param in your zf-mvc-auth config:

```php
  'locator_name'=>'ZF\OAuth2\Adapter\MongoAdapter',
```

in-situ : 

```php
 'zf-mvc-auth' => array(
        'authentication' => array(
            'adapters' => array(
                'oauth2_mongo' => array(
                    'adapter' => 'ZF\\MvcAuth\\Authentication\\OAuth2Adapter',
                    'storage' => array(
                        'adapter' => 'mongo',
                           'locator_name'=>'ZF\OAuth2\Adapter\MongoAdapter',
                        'route' => '/oauth'

                    ),
                ),
            ),
        ),
    )
```

To summarise, rather than providing mongo credentials in your zf-mvc-auth config, simply provide the name of an existing db locator. In this case  'ZF\OAuth2\Adapter\MongoAdapter' which is configured for zf-oauth2 module to function. FYI the config file is here:

https://github.com/zfcampus/zf-oauth2/blob/master/config/oauth2.local.php.dist

Hope this helps @weckx @codeaid  @ezimuel  @weierophinney

**Comment by weckx** [comment]
This works! However I set locator_name to a service that returns a MongoDB instance. Thanks @matwright
